devsearch/preprocessing/extract.py

```diff
@@ -0,0 +1,176 @@
+        should_clone: If should clone GitHub then clone else repo should be in target path
+
+    Returns:
+
+    """
+    if should_clone:
+        porcelain.clone(git_path, target_path)
+
+    repo = Repo(target_path)
+    repo_url = (repo.get_config().get((b"remote", b"origin"), b"url")).decode()
+
+    for entry in tqdm(repo.get_walker()):
+        entry: WalkEntry
+        commit: Commit = entry.commit
+
+        author = commit.author.decode()
+        commit_sha = commit.id.decode()
+
+        for change_sequence in entry.changes():
+            # for some reason change could be a sequence of changes - despite it is a single commit
+            if not isinstance(change_sequence, list):
+                change_sequence = [change_sequence]
+
+            for change in change_sequence:
+                file_name = change.new.path or change.old.path
+                file_name = file_name.decode()
+
+                # print(file_name)
+                # we process only some extensions todo
+                if not file_name.endswith(tuple(SUPPORTED_EXTENSIONS)):
+                    continue
+
+                path = f"{repo_url}/blob/{commit_sha}/{file_name}"
+                blob_id = change.new.sha or change.old.sha
+                blob_id = blob_id.decode()
+
+                new_entity = {"author": author,
+                              "commit_sha": commit_sha,
+                              "path": path,
+                              "repo_url": repo_url,
+                              "blob_id": blob_id}
+                new_entity.update(get_change_differences(repo, change))
+
+                yield new_entity
+
+
+def save_data_as_json(data: Union[List, Dict], path: str) -> None:
+    """
+    A method to serialize json-like object
+    Args:
+        data: data to save
+        path: path to save like C:/path/to/file.jsonl
+
+    """
+    assert path.endswith("jsonl")
+
+    with open(path, "a") as fp:
+        fp.write(json.dumps(data))
+        fp.write("\n")
+
+
+def process_stargazers(repo_name: str, max_user_stars: int = 200, top_repos: int = 100) -> Dict[str, int]:
+    """
+    A method that does following:
+    1. get all git repo stargazers (list)
+    2. for each stargazer get all of his starred repos
+    3. for each visited repo during previous step saves its stargazers count
+    4. returns mapping "repo" - "how much stars from our list repo has"
+    Args:
+        repo_name: path to github repo like "scikit-learn/scikit-learn"
+        max_user_stars: users with count of starred repos > max_user_stars - ignored
+        top_repos: returns this number of most popular repos
+
+    Returns: {"repo_name": "stargazers"}
+    """
+    repo_to_stars = Counter()
+    g = Github(os.getenv("access_token"))
+
+    repo = None
+    while repo is None:
+        try:
+            repo = g.get_repo(repo_name)
+        except RateLimitExceededException as e:
+            print(f"api limit - {e}")
+            time.sleep(1000)
```

Comment:
https://github.com/PyGithub/PyGithub/issues/1115
https://github.com/PyGithub/PyGithub/issues/1319#issuecomment-567557766
I think it should be possible to estimate when you can query next